### DISCUSSION TITLE
Why do we support both CMake and autotools?

### DISCUSSION THREAD

**Discussion by drwells** [IBAMR/IBAMR#1467] [General]
IBAMR presently supports configuration and compilation with two different build systems: autotools and CMake. In brief, the CMake-based build system will ultimately replace the autotools-based one due to its better feature set and lower overall maintenance cost. This discussion thread summarizes some FAQ and provides a place for people to discuss the new CMake build system.

### PETSc uses BuildSystem and not CMake - why don't we use BuildSystem?
- PETSc's build system needs to handle many, many things that we do not. For example: PETSc can work with dozens of optional dependencies (we have two optional dependencies) and can download, compile, and install both its dependencies and a large number of external projects. We simply do not need to cover this use case.
- Since we are a C++ project, in practice we only support three compiler toolchains: GCC, clang, and intel. PETSc supports using a larger variety of C compilers. This also lowers the complexity of our build system and makes using an 'off-the-shelf' one more reasonable.
- The build system we use must be something the principal IBAMR authors can maintain. There is substantial documentation online for using CMake. In comparison, to the best of @drwells' knowledge only PETSc uses BuildSystem and its documentation is combined with that of PETSc.

### Why CMake instead of autotools?
In general, comparing these two build systems is somewhat difficult since autotools significantly predates CMake and served as the inspiration for much of CMake's design. The problems autotools set out to solve defined the initial scope and methods for CMake.

- CMake can generate files which any IDE can understand.
- CMake scripts are written in a single language. A student will need to learn some CMake scripting, which is fairly well-documented and there are many, many online examples to look at. In comparison, adding new features to our autotools-based build system requires knowledge of shell, m4, and automake syntax: three languages instead of one.
- CMake tends to be much simpler. For example, to add new Fortran files in #1464 the author needed to edit eight autotools files and understand a lot of implicit knowledge about how we convert m4 into Fortran. By comparison, only one CMake script needs to be modified.
- We can skip using `autoreconf` and no longer have to deal with huge amounts of code changes resulting from increasing the version number of automake.
- CMake has a number of useful features that simply do not exist in autotools, such as its ability to detect unresolvable cyclic link dependencies (which are typically the result of a user error), better support for installation and exporting our dependencies, and support for sophisticated features like link-time optimization.
- The ability to express our dependencies as targets results in a vastly simpler and easier to understand build system and allows us to perform some useful checks, e.g., checking for compile or link time conflicts between our dependencies.
- CMake has better support for using shared object files and does not rely on scripts like `libtool` for relocating them.
- The CMake build system is much simpler than the autotools one. Aside from examples (which get individual `CMakeLists.txt` files), the build system consists of `CMakeLists.txt`, `ibtk/src/CMakeLists.txt`, `src/CMakeLists.txt`, `tests/CMakeLists.txt`, and `cmake/IBAMRConfig.cmake.in`: that's a total of five files for configuring the library, configuring the test suite, managing all dependencies, and installing information about our dependencies.

### Why autotools instead of CMake?
While moving to CMake is the right choice for us there are still many things which autotools does better:
- The generated `configure` script will run on essentially any Unix machine from the last 25 years. While we are not that portable due to our C++11 requirement, a major advantage of autotools is that the build system itself will work with any Unix machine.
- There is no requirement that the user install a recent copy of `CMake` or any other tool: everything is included in the `configure` script.
- CMake does not have an equivalent to `./configure --help`: we instead rely on information in the manual which may be out of date or difficult to locate for new users.
- CMake `FindPackage()` functions can be brittle or break in weird ways. As a result our detection logic for MPI is much more complex for CMake than autotools.

### How long will IBAMR support both CMake and autotools?
- A date has not yet been set. We will probably drop support for autotools when we require SAMRAI 4.

**Comment by drwells** [comment]
@amneetb if you have more questions let me know - it would be good to collect all of this information in one place.

**Reply by amneetb** [reply]
In autotools one can insert bash commands/scripts. Is there something similar in CMake? Maybe point out some useful resources here for users and developers to go over.

**Reply by amneetb** [reply]
Also there must be some advantages of autotools over CMake. It is hard to believe that there are only disadvantages. The GNU project is not that bad after all.

**Reply by drwells** [reply]
1. CMake itself is a scripting language. For example we check that SAMRAI has MPI with
```cmake
SET(CMAKE_REQUIRED_INCLUDES ${SAMRAI_INCLUDE_DIRS})
CHECK_CXX_SOURCE_COMPILES(
  "
  #include <SAMRAI_config.h>
  #ifndef HAVE_MPI
  #error
  #endif
  int main() {}
  "
  SAMRAI_WITH_MPI)
SET(CMAKE_REQUIRED_INCLUDES "")
IF(NOT ${SAMRAI_WITH_MPI})
  MESSAGE(FATAL_ERROR "IBAMR requires that SAMRAI is built with MPI support.")
ENDIF()
```
Unlike some newer build systems (e.g, meson) we still have a real programming language available to us for writing such checks.

2. I added some.

**Reply by amneetb** [reply]
Thanks!

**Reply by amneetb** [reply]
> @amneetb if you have more questions let me know - it would be good to collect all of this information in one place.


Getting error with CMake on OS X
```

amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/ -DBOOST_ROOT=/Users/amneetb/Softwares/boost/1.69.0 -DPETSC_ROOT=/Users/amneetb/Softwares/PETSc-Gitlab/PETSc/darwin-dbg/ -DMPI_ROOT=/Users/amneetb/Softwares/MPICH -DHYPRE_ROOT=/Users/amneetb/Softwares/PETSc-Gitlab/PETSc/darwin-dbg/ -DSAMRAI_ROOT=/Users/amneetb/Softwares/SAMRAI/darwin-dbg -DHDF5_ROOT=/Users/amneetb/Softwares/HDF5 -DSILO_ROOT=/Users/amneetb/Softwares/SILO -DLIBMESH_ROOT=/Users/amneetb/Softwares/libMesh-Git/darwin-dbg -DLIBMESH_METHOD=DBG ../IBAMR
-- This is CMake 3.21.1
-- 
-- Detecting Fortran/C Interface
-- Detecting Fortran/C Interface - Found GLOBAL and MODULE mangling
-- BOOST_ROOT=/Users/amneetb/Softwares/boost/1.69.0
-- EIGEN3_ROOT was not provided to CMake: default search paths will be used.
-- HDF5_ROOT=/Users/amneetb/Softwares/HDF5
-- HYPRE_ROOT=/Users/amneetb/Softwares/PETSc-Gitlab/PETSc/darwin-dbg/
-- MUPARSER_ROOT was not provided to CMake: default search paths will be used.
-- MPI_ROOT=/Users/amneetb/Softwares/MPICH
-- PETSC_ROOT=/Users/amneetb/Softwares/PETSc-Gitlab/PETSc/darwin-dbg/
-- SAMRAI_ROOT=/Users/amneetb/Softwares/SAMRAI/darwin-dbg
-- LIBMESH_ROOT=/Users/amneetb/Softwares/libMesh-Git/darwin-dbg
-- NUMDIFF_ROOT was not provided to CMake: since this is an optional dependency IBAMR will be configured without this package.
-- SILO_ROOT=/Users/amneetb/Softwares/SILO
-- 
-- Setting up MPI
-- CMAKE_CXX_COMPILER=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ is not an MPI compiler wrapper
-- Setting up MPI at location /Users/amneetb/Softwares/MPICH
-- Could NOT find MPI_Fortran (missing: MPI_Fortran_WORKS) 
CMake Error at /usr/local/Cellar/cmake/3.21.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find MPI (missing: MPI_Fortran_FOUND) (found version "3.1")
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.21.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.21.1/share/cmake/Modules/FindMPI.cmake:1747 (find_package_handle_standard_args)
  CMakeLists.txt:205 (FIND_PACKAGE)


-- Configuring incomplete, errors occurred!
See also "/Users/amneetb/Softwares/IBAMR-Git/cmake-objs-dbg/CMakeFiles/CMakeOutput.log".
See also "/Users/amneetb/Softwares/IBAMR-Git/cmake-objs-dbg/CMakeFiles/CMakeError.log".
```

**Reply by amneetb** [reply]
FYI

```
amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ ls /Users/amneetb/Softwares/MPICH/bin
hydra_nameserver* mpic++@           mpicxx*           mpif77@           mpirun@
hydra_persist*    mpicc*            mpiexec@          mpif90@           mpivars*
hydra_pmi_proxy*  mpichversion*     mpiexec.hydra*    mpifort*          parkill*

```

**Reply by drwells** [reply]
Can you upload `CMakeFiles/CMakeError.log` and `CMakeFiles/CMakeOutput.log` here?

**Comment by amneetb** [comment]
@drwells : Attached

[CMakeError.log](https://github.com/IBAMR/IBAMR/files/8921129/CMakeError.log)
[CMakeOutput.log](https://github.com/IBAMR/IBAMR/files/8921130/CMakeOutput.log)

**Comment by amneetb** [comment]
@drwells: It is also not able to find `ld` on OS X. This is captured during configure with a warning. However, during build time it leads to an error. I used your latest fix to skip checking MPI_Fortran to get past CMake configure error. 

```
[  9%] Building Fortran object ibtk/src/CMakeFiles/IBTK3d.dir/boundary/cf_interface/fortran/linearcfinterpolation3d.f.o
ld: framework not found -c
ld: framework not found -c
ld: framework not found -c
ld: framework not found -c
ld: framework not found -c
ld: framework not found -c
ld: framework not found -c
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/coarsen_ops/fortran/cubiccoarsen2d.f.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/math/fortran/rot2d.f.o] Error 1
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/math/fortran/graddetect2d.f.o] Error 1
make[2]: ld: framework not found -c
*** [ibtk/src/CMakeFiles/IBTK2d.dir/math/fortran/curl2d.f.o] Error 1
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/coarsen_ops/fortran/rt0coarsen2d.f.o] Error 1
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/boundary/cf_interface/fortran/linearcfinterpolation2d.f.o] Error 1
make[2]: *** [ibtk/src/CMakeFiles/IBTK3d.dir/boundary/cf_interface/fortran/linearcfinterpolation3d.f.o] Error 1
make[2]: *** Waiting for unfinished jobs....
collect2: error: ld returned 1 exit status
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/math/fortran/div2d.f.o] Error 1
ld: framework not found -c
collect2: error: ld returned 1 exit status
ld: framework not found -c
collect2: error: ld returned 1 exit status
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/math/fortran/miscmath2d.f.o] Error 1
make[2]: *** [ibtk/src/CMakeFiles/IBTK2d.dir/math/fortran/grad2d.f.o] Error 1

```

**Comment by amneetb** [comment]
It is surprising to see how brittle CMake is on some of the essential things like `ld` and compilers!

**Reply by drwells** [reply]
I'm not convinced that this is a CMake problem. Check out this compiler line:
```
/usr/local/bin/gfortran  -I/Users/amneetb/Softwares/MPICH/include -O3 -DNDEBUG -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -fPIE -framework -c /Users/amneetb/Softwares/IBAMR-Git/cmake-objs-dbg/CMakeFiles/FindMPI/test_mpi.f90 -o CMakeFiles/cmTC_a0934.dir/test_mpi.f90.o
```
This fails because there is no framework named `-c`. This line probably comes from the MPI compiler wrapper script. I'll see if I can reproduce this locally.

**Reply by amneetb** [reply]
If it helps

```
amneetb@Amneets-MacBook-Pro:~/Softwares/MPICH/bin$ ./mpichversion 
MPICH Version:    	3.4.2
MPICH Release date:	Wed May 26 15:51:40 CDT 2021
MPICH Device:    	ch4:ofi
MPICH configure: 	CC=clang CXX=clang++ FC=gfortran F77=gfortran --prefix=/Users/amneetb/Softwares/MPICH FFLAGS=-fallow-argument-mismatch
MPICH CC: 	clang    -O2
MPICH CXX: 	clang++   -O2
MPICH F77: 	gfortran -fallow-argument-mismatch  -O2
MPICH FC: 	gfortran   -O2
MPICH Custom Information: 

amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ clang --version
Apple clang version 12.0.5 (clang-1205.0.22.11)
Target: x86_64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin


amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ gfortran --version
GNU Fortran (Homebrew GCC 11.2.0) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ ldd --version
llvm-otool(1): Apple Inc. version cctools-980.1
otool(1): Apple Inc. version cctools-980.1
disassmbler: LLVM version 12.0.5, (clang-1205.0.22.11)


amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ ld -v
@(#)PROGRAM:ld  PROJECT:ld64-650.9
BUILD 13:09:02 May 28 2021
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 12.0.5, (clang-1205.0.22.11) (static support for 27, runtime is 27)
TAPI support using: Apple TAPI version 12.0.5 (tapi-1205.0.7.1)

amneetb@Amneets-MacBook-Pro:~/Softwares/IBAMR-Git/cmake-objs-dbg$ gfortran -v
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/11.2.0/libexec/gcc/x86_64-apple-darwin20/11.2.0/lto-wrapper
Target: x86_64-apple-darwin20
Configured with: ../configure --prefix=/usr/local/Cellar/gcc/11.2.0 --libdir=/usr/local/Cellar/gcc/11.2.0/lib/gcc/11 --disable-nls --enable-checking=release --enable-languages=c,c++,objc,obj-c++,fortran,d --program-suffix=-11 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-zstd=/usr/local/opt/zstd --with-pkgversion='Homebrew GCC 11.2.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --enable-libphobos --build=x86_64-apple-darwin20 --with-system-zlib --disable-multilib --without-build-config --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Homebrew GCC 11.2.0) 

```

**Comment by amneetb** [comment]
@drwells: Did you get a chance to look into CMake problems on MacOS?

**Reply by drwells** [reply]
Not yet - I need to actually recompile mpich and everything else in order to reproduce this.

There's one more quick thing we should check: can you paste the output of `mpifort -show` here?

**Comment by amneetb** [comment]
```
amneetb@Amneets-MacBook-Pro:~/Softwares/PythonCMAES$ mpifort -show
gfortran -framework OpenCL -Wl,-flat_namespace -Wl,-commons,use_dylibs -I/Users/amneetb/Softwares/MPICH/include -I/Users/amneetb/Softwares/MPICH/include -L/Users/amneetb/Softwares/MPICH/lib -lmpifort -lmpi -lpmpi

```

**Reply by drwells** [reply]
Thanks. It's interesting that `-framework OpenCL` got truncated to `-framework`. That's more likely to be an odd CMake problem.

**Reply by drwells** [reply]
Yup, its a bug with how CMake parses things:

https://gitlab.kitware.com/cmake/cmake/-/issues/21741

I'll see if I can figure out a workaround.

**Reply by drwells** [reply]
One workaround is to use the MPI compiler wrappers. That works on my machine while just setting `MPI_ROOT` fails.

**Reply by amneetb** [reply]
I can try that. Can you send me the exact command line you used?

**Reply by drwells** [reply]
sure. I installed hypre and hdf5 through petsc, so they are all in the same place:

```sh
#!/bin/bash
set -e

petsc=$HOME/Applications/petsc-3.13.1-mpich/x86_64-opt

MPI_ROOT=$HOME/Applications/mpich-4.0.2

cmake \
    -DCMAKE_CXX_COMPILER=$MPI_ROOT/bin/mpic++ \
    -DCMAKE_C_COMPILER=$MPI_ROOT/bin/mpicc \
    -DCMAKE_Fortran_COMPILER=$MPI_ROOT/bin/mpifort \
    -DPETSC_ROOT=$petsc \
    -DHDF5_ROOT=$petsc \
    -DHYPRE_ROOT=$petsc \
    -DSAMRAI_ROOT=$HOME/Applications/SAMRAI-2.4.4-mpich \
    ..

make -j8
```
I don't yet understand why this works but just providing MPI_ROOT does not. Regardless its a bug in cmake.

**Comment by amneetb** [comment]
All the more reasons to support autotools as well. FWIW, Rahul was able to build IBAMR on Fugaku and K-computer in Japan using autotools (which speaks to its strength).

**Reply by drwells** [reply]
Indeed - we have put a lot of work into the autotools build system and it does what we need it to do. It would be interesting to hear about an experience of someone cross-compiling IBAMR with CMake too, though :)

**Comment by amneetb** [comment]
Ok, the CMake works on IBAMR with the specified compilers. I was trying to define a `CMakeLists.txt` file for a project and I ran into the following errors during compilation:

```
amneetb@Amneets-MacBook-Pro:~/Softwares/CFD-MPC-Git$ make 
Consolidate compiler generated dependencies of target main3d
[ 10%] Building CXX object CMakeFiles/main3d.dir/main.cpp.o
clang: warning: -framework -std=gnu++11: 'linker' input unused [-Wunused-command-line-argument]
In file included from /Users/amneetb/Softwares/CFD-MPC-Git/main.cpp:34:
In file included from /Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibamr/AdvDiffSemiImplicitHierarchyIntegrator.h:23:
In file included from /Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibamr/AdvDiffHierarchyIntegrator.h:27:
In file included from /Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibtk/HierarchyIntegrator.h:24:
In file included from /Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibtk/HierarchyMathOps.h:24:
/Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibtk/SAMRAIDataCache.h:187:27: error: no template named 'tuple' in namespace 'std'
    using key_type = std::tuple<std::type_index, /*data_depth*/ int, /*ghost_cell_width*/ int>;
                     ~~~~~^
/Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibtk/SAMRAIDataCache.h:190:19: error: use of undeclared identifier 'key_type'
    std::multimap<key_type, int> d_available_data_idx_map, d_unavailable_data_idx_map;
                  ^
/Users/amneetb/Softwares/IBAMR-Git/cmake-install-dbg/include/ibtk/SAMRAIDataCache.h:196:12: error: unknown type name 'key_type'
    static key_type construct_data_descriptor(int idx);

```

**Reply by drwells** [reply]
Unfortunately that's the same issue manifested in a different way - `-framework -std=gnu++11` eats the C++ version flag.

Some possible fixes:
1. patch CMake (I am looking into it)
2. provide another explicit version flag via `-DCMAKE_CXX_FLAGS="-std=c++11"`
3. Delete `-framework OpenCL` from the list of flags in `mpicc`, `mpic++`, and `mpifort` - unless you want to use OpenCL you don't need it anyway

**Reply by boyceg** [reply]
I think you also could add the C++ version flag to the name of the C++ compiler --- something like:
```
    -DCMAKE_CXX_COMPILER="$MPI_ROOT/bin/mpic++ -std=c++11"
```

**Reply by amneetb** [reply]
I added the same compiler options (without std C++11) to my project CMake configure. This made it work.
